src/runner_manager.py

```diff
@@ -399,55 +506,19 @@ def reconcile(self, quantity: int, resources: VirtualMachineResources) -> int:
         delta = quantity - len(runner_states.healthy)
         # Spawn new runners
         if delta > 0:
-            if not RunnerManager.runner_bin_path.exists():
-                raise RunnerCreateError("Unable to create runner due to missing runner binary.")
-
-            logger.info("Getting registration token for GitHub runners.")
-
-            registration_token = self._get_github_registration_token()
-            remove_token = self._get_github_remove_token()
-
-            logger.info("Attempting to add %i runner(s).", delta)
-            for _ in range(delta):
-                config = RunnerConfig(
-                    self.app_name,
-                    self.config.path,
-                    self.proxies,
-                    self.config.lxd_storage_path,
-                    self._generate_runner_name(),
-                    self.config.charm_state.is_metrics_logging_available,
-                )
-                runner = Runner(self._clients, config, RunnerStatus())
-                try:
-                    self._create_runner(registration_token, resources, runner)
-                    logger.info("Created runner: %s", runner.config.name)
-                except RunnerCreateError:
-                    logger.error("Unable to create runner: %s", runner.config.name)
-                    runner.remove(remove_token)
-                    logger.info("Cleaned up runner: %s", runner.config.name)
-                    raise
-
+            self._spawn_new_runners(delta, resources)
         elif delta < 0:
-            logger.info("Attempting to remove %i runner(s).", -delta)
-            # Idle runners are online runners that have not taken a job.
-            idle_runners = [runner for runner in online_runners if not runner.status.busy]
-            offset = min(-delta, len(idle_runners))
-            if offset != 0:
-                logger.info("Removing %i runner(s).", offset)
-                remove_runners = idle_runners[:offset]
-
-                logger.info("Cleaning up idle runners.")
-
-                remove_token = self._get_github_remove_token()
-
-                for runner in remove_runners:
-                    runner.remove(remove_token)
-                    logger.info("Removed runner: %s", runner.config.name)
-            else:
-                logger.info("There are no idle runners to remove.")
+            self._remove_runners(count=-delta, runners=online_runners)
         else:
             logger.info("No changes to number of runners needed.")
 
+        if self.config.charm_state.is_metrics_logging_available:
+            end_ts = time.time()
+            self._issue_reconciliation_metric(
```

Comment:
It's not very clear to me: What does the issue_reconciliation_metric do?

Reply:
The`ìssue_reconciliation_metric` method creates the reconciliation metric log entry, which looks like this:
```json
{
"event": "reconciliation",
"log_timestamp": 1694687032,
"flavor": "small",
"crashed_runners": 10,
"idle_runners": 12,
"duration": 600,
}
```